Docsp 9562 geospatial queries (#78)
mongodb/docs-java: source/fundamentals/crud/read-operations/geo.txt

===================
Search Geospatially
===================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> On this page, you can learn how to query **geospatial data**

Author: Thanks for the note. Committing suggested change. 

This sentence was rewritten in Team PR review and I think Chris has a  preference for the "you can learn" phrasing. I personally don't have a preference between the phrasings, but I think we should reach consensus as a team on which phrasings are acceptable (as I understand it you prefer the "you will" phrasing and Chris prefers the "you can" phrasing in introductions). As it stands now, *no matter which one a writer picks it will be requested to be changed by someone on the team* (which seems incorrect). 

I am committing the suggested change, but I think we should reach consensus on this issue as a team and maybe make more rigid style guidelines for our introductions such that we have a clear definition of success for writers.

with the MongoDB Java Driver. 

Geospatial data is data tied to a geographical location on

> the surface of the Earth. Examples of geospatial data include:

Author: Thanks for the note. After rereading, I agree that "is tied to" is colloquial.  Committing suggested change. 

Going forward, I continue to try and refine my sense for colloquial language and avoid it in my writing. I wonder if we can compile a dictionary of colloquial phrases and have vale automate checks for it?

- Locations of movie theaters
- Borders of countries
- Routes of bicycle rides
- Dog exercise areas in New York City

> You should read this guide if you would like to:

Author: Thanks for the note. After rereading this a few times, I can see how it doesn't add much for readers. I rewrote the introductory sentence to try and communicate the same message as this list. Fix:

Responding to feedback around "should... if you would like to": This is the phrasing Chris and I arrived at after a few meetings around the [Sorts Page](https://docs-mongodbcom-staging.corp.mongodb.com/a54a90a/java/docsworker-xlarge/DOCSP-9562-Geo/fundamentals/builders/sort/) (first page I wrote that received high volume of notes and had a few synchronous meetings). If I remember correctly, this passive phrasing was considered desirable as we are discussing the reader's behavior (it could be rude to be assertive in saying "you must do"). I personally agreed with this stance of being passive when discussing the reader's behavior, but I am totally fine not using this phrasing.

I think this would be a good point of discussion in a team PR review meeting.

> You should read this guide if you would like to:

Author: In my final proofread, I ended up changing this to https://github.com/mongodb/docs-java/pull/78/commits/e897a9d9371c9c437a67590e7e4eab524684e834. I think changing the top heading to "search **geospatial data** leaves work for the list at the bottom of the intro to break down the concrete concepts we will cover on the page later on.

Let me know if you disagree, and as this is a comment that it not intuitive to me, it may be worth discussing it synchronously.

- Understand the different geospatial data formats in MongoDB.
- Build geospatial queries.

Coordinates on an Earth-like Sphere
-----------------------------------

You can represent geospatial data on an Earth-like sphere with the **GeoJSON**
format. GeoJSON is a geospatial data format. You can use the GeoJSON format to
store your data, and you must use the GeoJSON format to query your geospatial

> data with MongoDB. 

Reviewer: This paragraph feels a bit fragmented.

1. You can represent `geospatial data` on an Earth-like sphere with the `**GeoJSON** format`.
2. GeoJSON is a geospatial data format.
The only words that don't overlap are "is a". 

You can use [the] GeoJSON format to store your data, and ... This doesn't feel very prescriptive. Why not something like
> Store and query your geospatial data in MongoDB using the GeoJSON format.

Author: Thanks for the note. I agree that the first two sentence in this section are clunky and should be changed. I removed "format" from the first sentence, and let the second sentence do this work. I now introduce "IETF" in the second sentence.

Fix:

In regards to the second comment, I feel that "you can" ([based off of the Style Guide's definition of "can"](https://docs.mongodb.com/meta/style-guide/writing/use-effective-verbs/#use-modal-verbs-accurately))is appropriate as the page does not advise whether you should store your data as GeoJSON or Legacy coordinate pairs. By saying "store and query your data...", I believe we are implicitly stating that you should not use Legacy Coordinate Pairs (and if readers should not use legacy coordinate pairs than legacy coordinate pairs should be marked as deprecated).

> For a concrete reference, GeoJSON data often resembles the following structure:

Reviewer: question: at this point is there a distinction between "GeoJSON" and "GeoJSON data" that merits a higher word count?

Author: Great question. Whenever I was researching for this page, I had trouble understanding what was inside and outside the GeoJSON specification. I personally think saying "GeoJSON data" is helpful here as in my head it says, 

> "here is what a single piece of GeoJSON data looks like. This isn't the GeoJSON specification, but this might give you a sense of what the thing we are talking about will look like in your app".

Also saying "GeoJSON often resembles the following" feels incorrect to me, as GeoJSON has a precise definition, where as "GeoJSON data" can take many forms.

Please let me know if you disagree with this defense of saying "GeoJSON data".

.. code-block:: json

   "<field name>" : {
      "type": "<GeoJSON type>",
      "coordinates": [
         [<longitude>, <latitude>],
         ...
         [<longitude>, <latitude>]
      ]
   }

For definitive information on GeoJSON, see the

> `official IETF GeoJSON specification <https://tools.ietf.org/html/rfc7946>`__.

Reviewer: suggestion: drop "IETF GeoJSON". IETF isn't previously defined, and the subject (GeoJSON) is already established.

```suggestion
`official specification <https://tools.ietf.org/html/rfc7946>`__.
```

Reviewer 2: **suggestion:** You could also use :rfc:`official specification <7946>`.

Author: Thanks for the feedback Nathan and Tony.I didn't realize we had an IETF RFC directive, good to know. 

Chris C had an interesting piece of feedback in our team PR review stating that "the reader might interpret this specification as having something to do with MongoDB." I thought this was valid, as when researching this page it took me a little bit to understand that GeoJSON was not a MongoDB spec. I agree with Chris' feedback, and think it could benefit some readers to clarify in this link who made the specification (IETF).

I also agree with your feedback Nathan that the word "GeoJSON" is extraneous, but I support leaving IETF in. I have added a definition for IETF earlier in the section. Please let me know if you disagree with this assessment, still prefer I remove "IETF" after hearing this defense of keeping it in. 

Fix: https://github.com/mongodb/docs-java/pull/78/commits/7ea4ae1718fcaec2992efa934e276249ea390582

.. external resource

GeoJSON Types
~~~~~~~~~~~~~

> Your GeoJSON object's type determines its geometric shape.

Reviewer: Following with the available types might make it flow more naturally, and help define the shapes.

Author: Thanks for the note, after rereading this a few times I agree that going Type -> Position -> Type is awkward and hard to understand. I have moved "GeoJSON" position to its own section. Fix: 

Going forward, I will try to better get a sense for when a concept is progressing linearly vs. switching back and forth between something else.

> In GeoJSON, you make a geometric shape from **positions**. A position is the

Reviewer: This applies throughout, but I don't think repeating the subject is particularly helpful.

Author: Thanks for the note.

The rest of this comment is just background around my repeating context, feel free to ignore it if you are busy.

When I am learning programming concepts, I think one of the main challenges for me is keeping track of where I am. Am I in "geojson" world, am I in "mongodb" world, or am I in "java" world? I think my repetition of, "In <context>" is an attempt to alleviate this challenge for readers (not make reader exert energy keeping track of implied context). Whether it is effective or not I don't know. 

I would be super interested to see research around this. If we ever do another round of user research I would love to test different writing styles against comprehension, enjoyment, speed, etc...

Index
~~~~~

To query data stored as legacy coordinate pairs, you must add the field containing
legacy coordinate pairs to  a ``2d`` index. The following snippet creates a
``2d`` index on the ``coordinates`` field using the ``Indexes`` builder:

.. code-block:: java

   import static com.mongodb.client.model.Indexes.geo2d;

   // <MongoCollection setup code here>

   collection.createIndex(geo2d("coordinates"));

For more information on the ``Indexes`` builder, see our 
:doc:`guide on the Indexes builder </fundamentals/builders/indexes>`. 
   
For more information on legacy coordinate pairs, see the
:manual:`MongoDB server manual page on legacy coordinate pairs </geospatial-queries/#legacy-coordinate-pairs>`.

.. external resource

.. note::

   Spherical (``2dsphere``) and flat (``2d``) indexes support some, but
   not all, of the same query operators. For a full list of operators
   and their index compatibility, see the
   :manual:`manual entry for geospatial queries </geospatial-queries/#geospatial-query-operators>`.

   .. external resource

Geospatial Queries
------------------

Geospatial queries consist of query operators with GeoJSON query parameters.

Query Operators

> ~~~~~~~~~~~~~~~

Reviewer: It feels like this can be combined into "Geospatial Query Operators"

Author: Thanks for the note. After trying to rewrite this in line with your edit, I would like to defend the current structure. Defense:

- Current structure allows GeoJSON package to be discussed separately from the Filters class
- Current structure has side right TOC mirror structure of a query (query operator and parameter comprising a completed query). 

I think breaking this section into two h1 sections ("Geospatial Query Operators" and "Geospatial Query Parameters") would less explicitly communicate the relationship between these two ideas.

Let me know if I am missing the mark in regards to your feedback, and as this comment is not intuitive to me, I think I might benefit from discussing it synchronously in a short call or over slack.

===================
Search Geospatially
===================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> On this page you will learn the how to search **geospatial data** in the
> MongoDB Java Driver.
> 
> Geospatial data is data that represents a geographical location on
> the surface of the Earth. Examples of geospatial data include:
> 
> - Locations of movie theaters
> - Borders of countries
> - Routes of bicycle rides
> - Dog exercise areas in New York City
> 
> You should read this guide if you would like to:
> 
> - Understand the different geospatial data formats in MongoDB.
> - Build geospatial queries.

Reviewer: I still think this separates the purpose of the page into two sections, with a definition of geospatial data wedged in-between.

```suggestion
On this page you will learn the how to search **geospatial data** with the
MongoDB Java Driver, and the different geospatial data formats supported by MongoDB.

Geospatial data is data that represents a geographical location on
the surface of the Earth. Examples of geospatial data include:

- Locations of movie theaters
- Borders of countries
- Routes of bicycle rides
- Dog exercise areas in New York City
```

> Coordinates on an Earth-like Sphere

Reviewer: Here and throughout, I'd remove "earth-like". I don't believe MongoDB supports defining a custom center point, radius, equator, and meridian?

Author: I think this is correct, I can't find anything suggesting you can define your own sphere. This might be worth making a ticket for as the term "Earth-Like" seems to be used throughout most of our documentation.

https://docs.mongodb.com/manual/core/2dsphere/
https://docs.mongodb.com/manual/geospatial-queries/
https://docs.mongodb.com/drivers/node/current/fundamentals/crud/read-operations/geo/

Might be worth making a ticket to change these from "an earth-like sphere" to just "earth". 

Removed "earth-like sphere" from page:

-----------------------------------

> You can represent geospatial data on an Earth-like sphere with **GeoJSON**.
> GeoJSON is a geospatial data format created by the Internet Engineering
> Task Force (IETF). You can use the GeoJSON format to store your data in MongoDB,
> and you must use the GeoJSON format to query your geospatial data in MongoDB. 

Reviewer: I've given this more thought. The message here is mixed, still, for no good reason. "You can use x to store data, you must use x to query". I can't see any other outcome other than possible confusion over this. Why not spell it out? 

Also, the structure is still somewhat repetitive, and doesn't flow well. The bit about the IETF feels shoehorned in now. Here's an idea to work with.

```suggestion
You can store geospatial data on Earth with **GeoJSON**, a geospatial data format 
[created by the Internet EngineeringTask Force (IETF)(does this really add anything?)].

 MongoDB uses the :manual:`same reference system </reference/glossary/#std-term-WGS84>`
 as GPS satellites to calculate geospatial locations.
```

Author: Thanks for the note. I agree that my fist paragraph in this section is too repetitive, and I agree that "can... must" doesn't add anything and that saying this doesn't suggest that legacy coordinate pairs are deprecated (as they are meant to be on a 2d plane rather than earth). 

I rewrote this section to be more task oriented (starting with verbs "store" and "query"), and shortened the sentence that introduces IETF.  

I added the fact about satellites to the examples as a fun fact (as there is a picture of a map with a geometry on it).

Fix:

> For a concrete reference, GeoJSON data often resembles the following structure:

Reviewer 3: Is this how it's read/stored in MongoDB? If so, would it be helpful to focus on that information rather than describing it as "a concrete reference... often resembles"?

Author: Thanks for the note. In line with our discussion in team PR review, this was changed into the location of MongoDB headquarters. https://github.com/mongodb/docs-java/pull/78/commits/5f31b46a5700bb7bf4a78504a2b09b2b3253b629

I agree that the "concrete reference... often resembles" is not productive (and sort of nonsensical). I agree with your and Kailie's comments that just presenting the info is better here (and providing an actual concrete reference of an understandable place is better).

.. code-block:: json

   "<field name>" : {
      "type": "<GeoJSON type>",
      "coordinates": [
         [<longitude>, <latitude>],
         ...

>          [<longitude>, <latitude>]

Reviewer 3: The prior format showed that the longitude and latitude pairs were different by enumerating them. This might be confusing since it doesn't differentiate whether the placeholders contain the same values. E.g.
<longitude_1>, <lattitude_1>
...
<longitude_n>, <lattitude_n>

Author: Thanks for the note. I agree these should be enumerated. This ended up being removed in turning this into the "location of MongoDB headquarters" example here: https://github.com/mongodb/docs-java/pull/78/commits/28d9772c7463b4c18d3522eb444788775b0bde66

Going forward, I will try to enumerate placeholder lists from `1 to n`.

      ]
   }

For definitive information on GeoJSON, see the
:rfc:`official IETF specification <7946>`.

.. external resource

GeoJSON Positions
~~~~~~~~~~~~~~~~~

> A position is the base unit in GeoJSON, and represents a single place on Earth.
> A position is an array containing two or three number values: 
> 
> - a longitude in the first position
> - a latitude in the second position
> - an optional elevation in the third position 

Reviewer: suggestion: reduce repetition. Also, I think it's easy to misconstrue "single place" as "single point". Lastly, base unit doesn't strictly match the RFC, nor does it help the user work with GeoJSON in this article, I feel.

```suggestion
A position represents a location on Earth, and is comprised as 
an array containing two or three number values: 

- Longitude in the first position (required)
- Latitude in the second position (required)
- Elevation in the third position (optional)
```

Author: Thanks for the note. I agree that "base unit" is the incorrect term. The specification uses "fundamental geometry construct", which I agree is not the same thing as "base unit". I changed "is comprised of" to "exists in code as" to [avoid passive voice](https://dictionary.cambridge.org/us/grammar/british-grammar/consist-comprise-or-compose). 

Going forward, I will try to continue to refine my sense for when repetition is unhelpful.

I really like the edit you have for the list of array values! Having required/optional at the end is great, and removing a/an is great also.

Fix: https://github.com/mongodb/docs-java/pull/78/commits/28d9772c7463b4c18d3522eb444788775b0bde66

> .. tip:: Order Matters

Reviewer 3: Consider a less general title so that the reader can identify whether they need to read the tip.
When reading this title, I think it's straightforward to interpret as "does the order of the coordinates in the array matter?" when it's actually the ordering of the latitude and longitude coordinate pair

>   formats, as you may need to swap your longitude and latitude values. 

Author: Thank you for the note. After rereading this, I agree the warning paragraph can be removed.

Fix:

Responding to "but it reads as explanation first, warning second": I agree that is makes sense to warn then explain the warning. Seems like the explanation will be given more attention with that order. Will try to keep that in mind going forward.

~~~~~~~~~~~~~

Your GeoJSON object's type determines its geometric shape. Geometric shapes are
made up of positions.

> Here are some common GeoJSON types and their corresponding positions:

Author: Thanks for the note. I am not 100% sure I am understanding your comment, but I agree that "corresponding positions" is a poor choice of words.

Fix: https://github.com/mongodb/docs-java/pull/78/commits/4dc09e9f8a33b7a80a77c691e7556e387df689fa

> It could be confused with geo positions, and readers may not directly relate a position to a linestring or poly, for example

Not sure what specifically the means, are you saying that readers may have trouble seeing that GeoJSON positions can form more descriptive geometric structures? If this is a correct reading, does my edit provide a satisfactory fix?

  
  Be careful when moving geospatial data between GeoJSON and other
  formats, as you may need to swap your longitude and latitude values. 

GeoJSON Types
~~~~~~~~~~~~~

Your GeoJSON object's type determines its geometric shape. Geometric shapes are
made up of positions.

Here are some common GeoJSON types and their corresponding positions:

- ``Point``: a single position.
- ``LineString``: a list of two or more positions, thus forming a series of line segments.
- ``Polygon``: a list of positions in which the first and last

>   position are the same, thus enclosing some space.

Reviewer: How about some simple examples here? An intersection or address might be a point, a road or the Great Wall of China might be a LineString, city boundaries or the Vatican might be delineated with a Polygon.

Author: Thanks for the note. I agree this could be helpful (and enjoyable) for learners. I like the examples of the The Great Wall of China and The Vatican. I used the Location of a Sculpture for point, and link to the wikipedia article for the sculpture which prominently features its location. 

Fix:

Index
~~~~~

To query data stored in the GeoJSON format, add the field containing
GeoJSON data to a ``2dsphere`` index. The following snippet creates a
``2dsphere`` index on the ``location.geo`` field using the ``Indexes`` builder:

.. code-block:: java

   import static com.mongodb.client.model.Indexes.geo2dsphere;

   // <MongoCollection setup code here>

>    collection.createIndex(geo2dsphere("location.geo"));

Author: Thanks for the note. I agree the syntax proposed in team PR review by you and @kyuan-mongodb improves readability and is easier to copy and paste.

Fix: https://github.com/mongodb/docs-java/pull/78/commits/2504423c0a66649b7dd7dab0c241ee04ded3f97f

For more information on the ``Indexes`` builder, see our 
:doc:`guide on the Indexes builder </fundamentals/builders/indexes>`.

.. guide resource

Coordinates on a 2D Plane
-------------------------

You can store geospatial data using ``x`` and ``y`` coordinates on 
a two-dimensional Euclidean plane. We refer to coordinates on a two-dimensional
plane as "legacy coordinate pairs".

Legacy coordinate pairs have the following structure:

.. code-block:: json

   "<field name>" : [ x, y ]

Your field should contain an array of two values in which the first represents 
the ``x`` axis value and the second represents the ``y`` axis value.

Index
~~~~~

To query data stored as legacy coordinate pairs, you must add the field containing
legacy coordinate pairs to  a ``2d`` index. The following snippet creates a
``2d`` index on the ``coordinates`` field using the ``Indexes`` builder:

.. code-block:: java

   import static com.mongodb.client.model.Indexes.geo2d;

   // <MongoCollection setup code here>

>    collection.createIndex(geo2d("coordinates"));

Author: Link to previous comment: https://github.com/mongodb/docs-java/pull/78#discussion_r631160384

   :manual:`manual entry for geospatial queries </geospatial-queries/#geospatial-query-operators>`.

   .. external resource

Geospatial Queries
------------------

Geospatial queries consist of a query operator and GeoJSON shapes as query
parameters.

Query Operators
~~~~~~~~~~~~~~~

To query your geospatial data, use one of the following query operators: 

- ``$near``
- ``$geoWithin``
- ``$nearSphere``
- ``$geoIntersects`` *requires a 2dsphere index*

You can specify these query operators in the MongoDB Java driver with the
``near()``, ``geoWithin()``, ``nearSphere()``, and ``geoIntersects()`` utility
methods of the ``Filters`` builder class.

For more information on geospatial query operators, see the
:manual:`manual entry for geospatial queries </geospatial-queries/#geospatial-query-operators>`.

.. external resource

For more information on ``Filters``, see our 
:doc:`guide on the Filters builder </fundamentals/builders/filters>`.  

Query Parameters
~~~~~~~~~~~~~~~~

To specify the GeoJSON shapes used as parameters in geospatial queries, use the
``Position``, ``Point``, ``LineString``, and ``Polygon`` classes of the MongoDB
Java driver.

> 

Author: Thank you for the note. I agree this suggestion improves the readability of the sentence. I agree that my repetition of the term "parameter" here is not productive. Committing suggested change.

Going forward, I will try and continue to refine my sense for when repetition is productive and unproductive in my pages.

For a full list of GeoJSON shapes available in the MongoDB Java driver, see the
:java-docs:`GeoJSON package </apidocs/mongodb-driver-core/com/mongodb/client/model/geojson/package-summary.html>`.

.. external resource

Examples
--------

The following examples use the MongoDB Atlas sample dataset. You can learn how
to  set up your own free-tier Atlas cluster and how to load the sample dataset
in our :doc:`quick start guide </quick-start>`.

The examples use the ``theaters`` collection in the ``sample_mflix`` database

> from the sample dataset. The ``theaters`` collection contains a ``2dsphere`` index

Author: Thank you for the note. While I agree this is repetitive, I would like to defend leaving this here. As `theaters` and `sample_mflix` are arbitrary and undefined terms, I want to be extremely explicit with readers about where they are coming from and what importance readers should give them. 

I have read pieces of documentation and been horribly confused when a term was introduced that I thought was meaningful only to realize it was an arbitrary name for a piece of sample data in a supplementary file I had neglected to observe closely. While I think a reader could infer what these terms are referring to if they have a solid understanding of MongoDB terminology and Java code, I would rather be extremely explicit that these terms are just arbitrary names from a sample dataset.

If after hearing this defense you still feel this should be removed, please let me know and I will remove it.

on the ``location.geo`` field.

The examples require the following imports:

.. literalinclude:: /includes/fundamentals/code-snippets/Geo.java
   :language: java
   :dedent:
   :start-after: begin exampleImports
   :end-before: end exampleImports

You can find the  
`source code for the examples on Github here <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/Geo.java>`__.

.. external resource

Query by Proximity
~~~~~~~~~~~~~~~~~~

The ``$near`` query operator accepts a set of longitude-latitude coordinates and returns
documents ordered from nearest to farthest. In the example below, we use the
``near()`` static utility method of the ``Filters`` builder class to construct a

> ``$near`` query.

Author: Thanks for the note. I agree this switch to "here is how MongoDB works" from "to do a task, use this tool" is unproductive for readers. Here is the fix: 

 I have tried to capture the task oriented nature of your edit while removing the duplicate description of the near method present in the suggested change. Let me know if my edit misses the mark.

   {"location": {"address": {"city": "Bronx"}}}
   {"location": {"address": {"city": "New York"}}}
   {"location": {"address": {"city": "New York"}}}
   {"location": {"address": {"city": "Long Island City"}}}
   {"location": {"address": {"city": "New York"}}}
   {"location": {"address": {"city": "Secaucus"}}}
   {"location": {"address": {"city": "Jersey City"}}}
   {"location": {"address": {"city": "Elmhurst"}}}
   {"location": {"address": {"city": "Flushing"}}}
   {"location": {"address": {"city": "Flushing"}}}
   {"location": {"address": {"city": "Flushing"}}}
   {"location": {"address": {"city": "Elmhurst"}}}

.. note:: Fun Fact

   MongoDB uses the
   :manual:`same reference system </reference/glossary/#std-term-WGS84>`
   as GPS satellites to calculate geometries over the Earth.

For more information on the ``$near`` operator, see  
:manual:`the reference documentation for $near </reference/operator/query/near/#mongodb-query-op.-near>`.

For more information on ``Filters``,
:doc:`see our guide on the Filters builder </fundamentals/builders/filters>`.

Query Within a Range
~~~~~~~~~~~~~~~~~~~~

The ``$geoWithin`` operator selects documents that have geospatial data within a

> specified shape. 

Author: Thank you for the note. I have rewritten this example to use a more imperative/task-oriented structure in line with https://github.com/mongodb/docs-java/pull/78#discussion_r630218465. Please let me know if missed the mark.

Fix:
